### PR TITLE
Semi-supervised multi-layer GCN model for node classification - s4628887

### PR DESCRIPTION
**Pull request by uly6mlv**
Hi Shakes,

Please review my pull request containing an implementation for solving the node classification of the Facebook Large Page-Page Network dataset using a semi-supervised multi-layer GCN model. The model is trained for 200 epochs and has an overall accuracy of 0.92. 

**graph.py:** Containing functions to compute self-loop adjacency matrix from edges and normalise the adjacency matrix by D^{\frac{1}{2}}AD^{\frac{1}{2}}, where D is the degree matrix.

**GCNConv.py**:  Defining a GCN layer in tensorflow. 

**model.py**:  Building a multi-layer GCN model

**GCN_Node_Classification.ipynb**: Driver script for pre-processing the data, building and training the model, retrieving results, and evaluating the performance.

**README.md**: Documentation for the algorithm description and implementation.

Regards,
Ulysses

### CODE REVIEW

**Comment by gayanku** [comment]
- Submitted Oct 21
- Readme is well done with background, results, diagrams. Dont use latex in markdown : github markdown does not support latex formulas. Also much better to show the classification report for the test set (rather than the whole graph). Further, markdown support tables as well which can be used to present results, rather than as an image.
- multi layer (3) layer GCN, Good accuracy (+90%), Semi supervised split.
- Good code structure and modularization
- **Valid for NORMAL**

- GCN layer code is from https://github.com/danielegrattarola/spektral/blob/master/spektral/layers/convolutional/general_conv.py. This is ok but need to mention source and/or cite.
- There is no .py driver script. Code is in well documented notebook with layers and model separate.

**Comment by uly6mlv** [comment]
> - Submitted Oct 21
> 
> - Readme is well done with background, results, diagrams. Dont use latex in markdown : github markdown does not support latex formulas. Also much better to show the classification report for the test set (rather than the whole graph). Further, markdown support tables as well which can be used to present results, rather than as an image.
> 
> - multi layer (3) layer GCN, Good accuracy (+90%), Semi supervised split.
> 
> - Good code structure and modularization
> 
> - **Valid for NORMAL**
> 
> 
> 
> - GCN layer code is from https://github.com/danielegrattarola/spektral/blob/master/spektral/layers/convolutional/general_conv.py. This is ok but need to mention source and/or cite.
> 
> - There is no .py driver script. Code is in well documented notebook with layers and model separate.

The GCN layer code is similar to spektral, but I've created a class to define the GCN layer which is not a subclass to another class. Moreover my class doesn't contain other unused methods.

I've asked another tutor to confirm that it's OK to use a Jupyter Notebook file as a driver because I think it is a better for data visualisation.

Lastly, I can see that GitHub doesn't support latex, so is there any way to type a formula?

**Comment by shakes76** [comment]
## TF/Torch Usage
Adequate use and implementation
## Good Practice (Design/Commenting)
Good spacing and comments
Header blocks not present -1
## Algorithm 
Driver Script present
Shows Usage & Demo & Visualisation & Data usage 
Module present
No Data leakage
Difficulty: Normal -5
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch)
No feedback required
Request Description good
